### PR TITLE
[AIRAVATA-2719] Refactoring Replica Catalog Implementation

### PR DESCRIPTION
**Pull request by tilaks26**
Implementation of DataProductRepository and DataReplicaLocationRepository instead of the existing implementation class dependency.

Changes made -

Added the necessary entity classes.
Included all the methods for the 2 repository classes.
Included tests for the repository classes.
Made necessary changes to the RegistryServerHandler.java class.

### CODE REVIEW

File: modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/utils/JPAUtil/RepCatalogJPAUtils.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private final static Logger logger = LoggerFactory.getLogger(RepCatalogJPAUtils.class);
+
+    private static final String PERSISTENCE_UNIT_NAME = "replicacatalog_data";
+    private static final String REPLICACATALOG_JDBC_DRIVER = "replicacatalog.jdbc.driver";
+    private static final String REPLICACATALOG_JDBC_URL = "replicacatalog.jdbc.url";
+    private static final String REPLICACATALOG_JDBC_USER = "replicacatalog.jdbc.user";
+    private static final String REPLICACATALOG_JDBC_PWD = "replicacatalog.jdbc.password";
+    private static final String REPLICACATALOG_VALIDATION_QUERY = "replicacatalog.validationQuery";
+
+    @PersistenceUnit(unitName="replicacatalog_data")
```
Comment (machristie) [inline modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/utils/JPAUtil/RepCatalogJPAUtils.java]: Can you use `PERSISTENCE_UNIT_NAME` here and on line 44?

File: modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/utils/QueryConstants.java [hunk 2]
```diff
@@ -37,4 +38,10 @@
             "WHERE BQ.groupResourceProfileId LIKE : " + DBConstants.GroupResourceProfile.GROUP_RESOURCE_PROFILE_ID;
     String FIND_ALL_GROUP_COMPUTE_RESOURCE_POLICY = "SELECT CR FROM "+ ComputeResourcePolicyEntity.class.getSimpleName() + " CR " +
             "WHERE CR.groupResourceProfileId LIKE : " + DBConstants.GroupResourceProfile.GROUP_RESOURCE_PROFILE_ID;
+
+    String FIND_ALL_CHILD_DATA_PRODUCTS = "SELECT DP FROM " + DataProductEntity.class.getSimpleName() + " DP " +
+            "WHERE DP.parentProductUri LIKE: " + DBConstants.DataProduct.PARENT_PRODUCT_URI;
```
Comment (machristie) [inline modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/utils/QueryConstants.java]: Shouldn't it be LIKE followed by a space followed by a colon? Same on 45 and 46.

**Comment by machristie** [comment]
@tilaks26 you mentioned that you're getting a NullPointerException in RepCatalogJPAUtils, can you post the stack trace? I don't see anything obviously wrong in there so I'm not sure what the problem could be.

File: modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/repositories/replicacatalog/RepCatAbstractRepository.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.persistence.EntityManager;
+import javax.persistence.Query;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public abstract class RepCatAbstractRepository<T, E, Id> {
```
[the remark this answers is not in the document]
Comment (tilaks26) [reply modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/repositories/replicacatalog/RepCatAbstractRepository.java]: For the execute method, we'll have to specify which catalog's entity manager is to be used. How do we differentiate that?

File: modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/entities/replicacatalog/DataProductEntity.java [hunk 4]
```diff
@@ -20,24 +20,61 @@
...
 
-    @Id
-    @Column(name = "PRODUCT_URI")
+    @Column(name = "PRODUCT_TYPE")
+    private DataProductType dataProductType;
+
+    @ElementCollection(fetch = FetchType.EAGER)
+    @CollectionTable(name="DATA_PRODUCT_METADATA", joinColumns = @JoinColumn(name="PRODUCT_URI"))
+    @Column(name = "METADATA_KEY")
+    private Map<String, String> productMetadata;
```
[the remark this answers is not in the document]
Comment (tilaks26) [reply modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/entities/replicacatalog/DataProductEntity.java]: You mean like this?

```
@ElementCollection(fetch = FetchType.EAGER)
@CollectionTable(name="DATA_PRODUCT_METADATA", joinColumns = @JoinColumn(name="PRODUCT_URI"))
@MapKeyColumn(name = "METADATA_KEY")
@Column(name = "METADATA_VALUE")
private Map<String, String> productMetadata;
```

File: modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/repositories/replicacatalog/RepCatAbstractRepository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (machristie) [reply modules/registry-refactoring/src/main/java/org/apache/airavata/registry/core/repositories/replicacatalog/RepCatAbstractRepository.java]: We talked about this but I'll just comment here as well.

I was thinking we could have an AbstractRepository class that classes like RepCatAbstractRepository would extends from. AbstractRepository would implement all of the methods but also have an `abstract` `getEntityManager()` method that subclasses would implement.

Mentioning @sachinkariyattin to get his input as well.

**Comment by machristie** [comment]
@tilaks26 I merged PR #167 and I think that caused merge conflicts for this PR. Can you resolve the merge conflicts?

**Comment by tilaks26** [comment]
@machristie Thank you for the review. I have resolved the merge conflicts and made the requested changes.

**Pull request merged**